File: ios/Sources/CapacitorSqlitePlugin/CapacitorSqlitePlugin.swift
import Foundation
import Capacitor

// swiftlint:disable type_body_length
@objc(CapacitorSqlitePlugin)
public class CapacitorSqlitePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorSqlitePlugin"
    public let jsName = "CapacitorSqlite"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getPluginPlatform", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "close", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isOpen", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getVersion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSchemaVersion", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "vacuum", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "execute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "run", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "runBatch", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "runMany", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "query", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "beginTransaction", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "commitTransaction", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "rollbackTransaction", returnType: CAPPluginReturnPromise)
    ]

    let impl = CapacitorSqlite()
    let workQueue = DispatchQueue(label: "com.devioarts.capacitor.sqlite.plugin", qos: .userInitiated)

    deinit {
        // Deinitialization may happen on the main thread. Queue cleanup behind any
        // already-submitted plugin work instead of blocking deinit on SQLite close/WAL I/O.
        let implementation = impl
        workQueue.async {
            implementation.closeAll()
        }
    }

    // MARK: - getPluginPlatform

    @objc func getPluginPlatform(_ call: CAPPluginCall) {
        success(call, data: ["platform": "ios"])
    }

    // MARK: - isAvailable

    @objc func isAvailable(_ call: CAPPluginCall) {
        success(call, data: ["available": impl.isAvailable()])
    }

    // MARK: - open

    @objc func open(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "open")
            return
        }
        let readonly   = call.getBool("readonly") ?? false
        let directory  = call.getString("directory")
        let migrations: [[String: Any]]
        if call.options["migrations"] != nil {
            guard let rawMigrations = call.getArray("migrations") else {
                failure(call, code: "MIGRATION_FAILED", message: "'migrations' must be an array", method: "open")
                return
            }
            guard let parsedMigrations = rawMigrations as? [[String: Any]] else {
                failure(call, code: "MIGRATION_FAILED", message: "'migrations' must be an array of objects", method: "open")
                return
            }
            migrations = parsedMigrations
        } else {
            migrations = []
        }

        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                try self.impl.open(database: database, readonly: readonly, migrations: migrations, directory: directory)
                self.success(call)
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "open")
            } catch {
                self.failure(call, code: "OPEN_FAILED", message: "open: \(error.localizedDescription)", method: "open")
            }
        }
    }

    // MARK: - close

    @objc func close(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "close")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                try self.impl.close(database: database)
                self.success(call)
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "close")
            } catch {
                self.failure(call, code: "CLOSE_FAILED", message: "close: \(error.localizedDescription)", method: "close")
            }
        }
    }

    // MARK: - isOpen

    @objc func isOpen(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "isOpen")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            self.success(call, data: ["open": self.impl.isOpen(database: database)])
        }
    }

    // MARK: - getVersion

    @objc func getVersion(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "getVersion")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                let version = try self.impl.getVersion(database: database)
                self.success(call, data: ["version": version])
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "getVersion")
            } catch {
                self.failure(call, code: "VERSION_FAILED", message: "getVersion: \(error.localizedDescription)", method: "getVersion")
            }
        }
    }

    // MARK: - getSchemaVersion

    @objc func getSchemaVersion(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "getSchemaVersion")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                let version = try self.impl.getSchemaVersion(database: database)
                self.success(call, data: ["version": version])
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(
                    call,
                    code: code,
                    message: msg,
                    method: "getSchemaVersion"
                )
            } catch {
                self.failure(
                    call,
                    code: "SCHEMA_VERSION_FAILED",
                    message: "getSchemaVersion: \(error.localizedDescription)",
                    method: "getSchemaVersion"
                )
            }
        }
    }

    // MARK: - vacuum

    @objc func vacuum(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "vacuum")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                try self.impl.vacuum(database: database)
                self.success(call)
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "vacuum")
            } catch {
                self.failure(call, code: "VACUUM_FAILED", message: "vacuum: \(error.localizedDescription)", method: "vacuum")
            }
        }
    }

    // MARK: - execute

    @objc func execute(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "execute")
            return
        }
        guard let statements = call.getArray("statements") as? [String], !statements.isEmpty else {
            failure(call, code: "INVALID_PARAMS", message: "'statements' must be a non-empty [String]", method: "execute")
            return
        }
        guard statements.allSatisfy({ !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }) else {
            failure(call, code: "INVALID_PARAMS", message: "'statements' entries must be non-empty strings", method: "execute")
            return
        }
        let transaction = call.getBool("transaction") ?? true
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                let changes = try self.impl.execute(database: database, statements: statements, transaction: transaction)
                self.success(call, data: ["changes": changes])
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "execute")
            } catch {
                self.failure(call, code: "EXECUTE_FAILED", message: "execute: \(error.localizedDescription)", method: "execute")
            }
        }
    }

    // MARK: - run

    @objc func run(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "run")
            return
        }
        guard let statement = call.getString("statement"), !statement.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            failure(call, code: "INVALID_PARAMS", message: "'statement' is required", method: "run")
            return
        }
        let values: [Any]
        do {
            values = try NativeBridgeValues.decode(call.getArray("values") ?? [], label: "values")
        } catch let error as NativeBridgeValueError {
            failure(call, code: "INVALID_PARAMS", message: error.message, method: "run")
            return
        } catch {
            failure(call, code: "INVALID_PARAMS", message: "Invalid values", method: "run")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                let result = try self.impl.run(database: database, statement: statement, values: values)
                self.success(call, data: ["changes": result.changes, "lastInsertId": result.lastInsertId])
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "run")
            } catch {
                self.failure(call, code: "EXECUTE_FAILED", message: "run: \(error.localizedDescription)", method: "run")
            }
        }
    }

    // MARK: - runBatch

    @objc func runBatch(_ call: CAPPluginCall) {
        let nativeStart = DispatchTime.now().uptimeNanoseconds
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "runBatch")
            return
        }
        let getArrayStart = DispatchTime.now().uptimeNanoseconds
        guard let set = call.getArray("set") as? [[String: Any]], !set.isEmpty else {
            failure(call, code: "INVALID_PARAMS", message: "'set' must be a non-empty array of {statement, values?}", method: "runBatch")
            return
        }
        let getArrayEnd = DispatchTime.now().uptimeNanoseconds
        let transaction = call.getBool("transaction") ?? true
        let includeDiagnostics = call.getBool("__diagnostics") == true
        let diagnostics = includeDiagnostics ? BatchDiagnostics() : nil
        diagnostics?.set("pluginGetArrayMs", nanos: getArrayEnd - getArrayStart)
        // runBatch() decodes tagged BLOB envelopes while normalizing values inside
        // Database.parseBatch(). Avoid a second 10k-item native pass over the same
        // payload before enqueueing the actual SQLite work.
        diagnostics?.set("bridgeDecodeMs", nanos: 0)
        let scheduledAt = DispatchTime.now().uptimeNanoseconds
        executeSqlite { [weak self] in
            guard let self = self else { return }
            diagnostics?.set("queueWaitMs", start: scheduledAt)
            do {
                let result = try self.impl.runBatch(
                    database: database,
                    set: set,
                    transaction: transaction,
                    diagnostics: diagnostics
                )
                diagnostics?.set("nativeTotalMs", start: nativeStart)
                var data: [String: Any] = ["changes": result.changes, "lastInsertId": result.lastInsertId]
                if let diagnostics {
                    data["timings"] = diagnostics.timings
                }
                self.success(call, data: data)
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "runBatch")
            } catch {
                self.failure(call, code: "EXECUTE_FAILED", message: "runBatch: \(error.localizedDescription)", method: "runBatch")
            }
        }
    }

    // MARK: - beginTransaction

    @objc func beginTransaction(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "beginTransaction")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                try self.impl.beginTransaction(database: database)
                self.success(call)
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "beginTransaction")
            } catch {
                self.failure(call, code: "TRANSACTION_FAILED", message: "beginTransaction: \(error.localizedDescription)", method: "beginTransaction")
            }
        }
    }

    // MARK: - commitTransaction

    @objc func commitTransaction(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "commitTransaction")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                try self.impl.commitTransaction(database: database)
                self.success(call)
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "commitTransaction")
            } catch {
                self.failure(call, code: "TRANSACTION_FAILED", message: "commitTransaction: \(error.localizedDescription)", method: "commitTransaction")
            }
        }
    }

    // MARK: - rollbackTransaction

    @objc func rollbackTransaction(_ call: CAPPluginCall) {
        guard let database = call.getString("database") else {
            failure(call, code: "INVALID_PARAMS", message: "'database' is required", method: "rollbackTransaction")
            return
        }
        executeSqlite { [weak self] in
            guard let self = self else { return }
            do {
                try self.impl.rollbackTransaction(database: database)
                self.success(call)
            } catch CapacitorSqliteError.failed(let code, let msg) {
                self.failure(call, code: code, message: msg, method: "rollbackTransaction")
            } catch {
                self.failure(call, code: "TRANSACTION_FAILED", message: "rollbackTransaction: \(error.localizedDescription)", method: "rollbackTransaction")
            }
        }
    }

}
// swiftlint:enable type_body_length

File: android/src/main/kotlin/com/devioarts/capacitor/sqlite/CapacitorSqlitePlugin.kt
package com.devioarts.capacitor.sqlite

import android.util.Base64
import java.nio.charset.StandardCharsets
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import org.json.JSONArray
import org.json.JSONObject

@CapacitorPlugin(name = "CapacitorSqlite")
class CapacitorSqlitePlugin : Plugin() {

    private companion object {
        const val NATIVE_BLOB_BASE64_KEY = "__capacitorSqliteBlobBase64"
    }

    private lateinit var impl: CapacitorSqlite
    private val sqliteExecutor = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "capacitor-sqlite")
    }

    override fun load() {
        impl = CapacitorSqlite(context)
    }

    override fun handleOnDestroy() {
        if (::impl.isInitialized) {
            // Enqueue on the SQLite thread instead of calling impl.closeAll() directly:
            // this runs after any already-queued operations finish (in order, same as
            // every other method call), and keeps the close()/WAL-checkpoint I/O off the
            // calling (bridge/UI) thread.
            try {
                sqliteExecutor.execute { impl.closeAll() }
            } catch (_: RejectedExecutionException) {
                // handleOnDestroy() called more than once — nothing left to flush.
            }
        }
        // shutdown() (not shutdownNow()) lets already-queued PluginCalls — including the
        // closeAll() task just submitted above — run to completion instead of being
        // silently dropped without ever resolving their JS Promise. Do not wait here:
        // Capacitor may call handleOnDestroy() on the UI thread, and a large queued batch
        // or WAL checkpoint must not freeze that thread for up to several seconds.
        sqliteExecutor.shutdown()
    }

    // MARK: - Unified response helpers

    private fun success(call: PluginCall, data: JSObject = JSObject()) {
        call.resolve(JSObject().put("success", true).put("data", data))
    }

    private fun failure(call: PluginCall, code: String, message: String, method: String) {
        val details = JSObject()
            .put("nativeCode", code)
            .put("nativeMessage", message)
            .put("source", "android-native")
        call.resolve(
            JSObject()
                .put("success", false)
                .put("error", JSObject()
                    .put("code", code)
                    .put("message", message)
                    .put("platform", "android")
                    .put("method", method)
                    .put("details", details)
                )
        )
    }

    private fun errorCode(e: Exception, fallback: String): String {
        return if (e is CapacitorSqliteException) e.code else fallback
    }

    private fun executeSqlite(block: () -> Unit) {
        sqliteExecutor.execute(block)
    }

    private fun timingsToJson(timings: Map<String, Double>): JSObject {
        val out = JSObject()
        timings.forEach { (key, value) -> out.put(key, value) }
        return out
    }

    // MARK: - getPlatform

    @PluginMethod
    fun getPlatform(call: PluginCall) {
        success(call, JSObject().put("platform", "android"))
    }

    // MARK: - isAvailable

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        success(call, JSObject().put("available", impl.isAvailable()))
    }

    // MARK: - open

    @PluginMethod
    fun open(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "open")
        val readonly = call.getBoolean("readonly", false) ?: false
        val directory = call.getString("directory")
        val migrations = try {
            jsonArrayToListOfMaps(call.getArray("migrations"), "migrations")
        } catch (e: IllegalArgumentException) {
            return failure(call, "MIGRATION_FAILED", e.message ?: "Invalid migrations", "open")
        }

        executeSqlite {
            try {
                impl.open(database, readonly, directory, migrations)
                success(call)
            } catch (e: Exception) {
                failure(call, errorCode(e, "OPEN_FAILED"), e.message ?: "open failed", "open")
            }
        }
    }

    // MARK: - close

    @PluginMethod
    fun close(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "close")

        executeSqlite {
            try {
                impl.close(database)
                success(call)
            } catch (e: Exception) {
                failure(call, errorCode(e, "CLOSE_FAILED"), e.message ?: "close failed", "close")
            }
        }
    }

    // MARK: - isOpen

    @PluginMethod
    fun isOpen(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "isOpen")

        executeSqlite {
            try {
                success(call, JSObject().put("open", impl.isOpen(database)))
            } catch (e: Exception) {
                failure(call, errorCode(e, "UNKNOWN"), e.message ?: "isOpen failed", "isOpen")
            }
        }
    }

    // MARK: - getVersion

    @PluginMethod
    fun getVersion(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "getVersion")

        executeSqlite {
            try {
                val version = impl.getVersion(database)
                success(call, JSObject().put("version", version))
            } catch (e: Exception) {
                failure(call, errorCode(e, "VERSION_FAILED"), e.message ?: "getVersion failed", "getVersion")
            }
        }
    }

    // MARK: - getSchemaVersion

    @PluginMethod
    fun getSchemaVersion(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "getSchemaVersion")

        executeSqlite {
            try {
                val version = impl.getSchemaVersion(database)
                success(call, JSObject().put("version", version))
            } catch (e: Exception) {
                failure(
                    call,
                    errorCode(e, "SCHEMA_VERSION_FAILED"),
                    e.message ?: "getSchemaVersion failed",
                    "getSchemaVersion"
                )
            }
        }
    }

    // MARK: - vacuum

    @PluginMethod
    fun vacuum(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "vacuum")

        executeSqlite {
            try {
                impl.vacuum(database)
                success(call)
            } catch (e: Exception) {
                failure(call, errorCode(e, "VACUUM_FAILED"), e.message ?: "vacuum failed", "vacuum")
            }
        }
    }

    // MARK: - execute

    @PluginMethod
    fun execute(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "execute")
        val stmts = try {
            jsonArrayToStringList(call.getArray("statements"), "statements")
        } catch (e: IllegalArgumentException) {
            return failure(call, "INVALID_PARAMS", e.message ?: "Invalid statements", "execute")
        }
        if (stmts.isEmpty()) {
            return failure(call, "INVALID_PARAMS", "'statements' must be a non-empty [string]", "execute")
        }
        val transaction = call.getBoolean("transaction", true) ?: true

        executeSqlite {
            try {
                val changes = impl.execute(database, stmts, transaction)
                success(call, JSObject().put("changes", changes))
            } catch (e: Exception) {
                failure(call, errorCode(e, "EXECUTE_FAILED"), e.message ?: "execute failed", "execute")
            }
        }
    }

    // MARK: - run

    @PluginMethod
    fun run(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "run")
        val statement = call.getString("statement")
            ?: return failure(call, "INVALID_PARAMS", "'statement' is required", "run")
        if (statement.trim().isEmpty()) {
            return failure(call, "INVALID_PARAMS", "'statement' is required", "run")
        }
        val values = try {
            jsonArrayToValueList(call.getArray("values"), "values")
        } catch (e: IllegalArgumentException) {
            return failure(call, "INVALID_PARAMS", e.message ?: "Invalid values", "run")
        }

        executeSqlite {
            try {
                val result = impl.run(database, statement, values)
                success(call, JSObject().put("changes", result.changes).put("lastInsertId", result.lastInsertId))
            } catch (e: Exception) {
                failure(call, errorCode(e, "EXECUTE_FAILED"), e.message ?: "run failed", "run")
            }
        }
    }

    // MARK: - runBatch

    @PluginMethod
    fun runBatch(call: PluginCall) {
        val nativeStart = System.nanoTime()
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "runBatch")
        val getArrayStart = System.nanoTime()
        val rawSet = call.getArray("set")
        val getArrayMs = (System.nanoTime() - getArrayStart) / 1_000_000.0
        val decodeStart = System.nanoTime()
        val set = try {
            jsonArrayToListOfMaps(rawSet, "set")
        } catch (e: IllegalArgumentException) {
            return failure(call, "INVALID_PARAMS", e.message ?: "Invalid set", "runBatch")
        }
        val decodeMs = (System.nanoTime() - decodeStart) / 1_000_000.0
        if (set.isEmpty()) {
            return failure(call, "INVALID_PARAMS", "'set' must be a non-empty array", "runBatch")
        }
        val transaction = call.getBoolean("transaction", true) ?: true
        val includeDiagnostics = call.getBoolean("__diagnostics", false) ?: false
        val timings = if (includeDiagnostics) linkedMapOf<String, Double>() else null
        timings?.put("pluginGetArrayMs", getArrayMs)
        timings?.put("bridgeDecodeMs", decodeMs)

        val scheduledAt = System.nanoTime()
        executeSqlite {
            timings?.put("queueWaitMs", (System.nanoTime() - scheduledAt) / 1_000_000.0)
            try {
                val result = impl.runBatch(database, set, transaction, timings)
                timings?.put("nativeTotalMs", (System.nanoTime() - nativeStart) / 1_000_000.0)
                val data = JSObject().put("changes", result.changes).put("lastInsertId", result.lastInsertId)
                timings?.let { data.put("timings", timingsToJson(it)) }
                success(call, data)
            } catch (e: Exception) {
                failure(call, errorCode(e, "EXECUTE_FAILED"), e.message ?: "runBatch failed", "runBatch")
            }
        }
    }

    // MARK: - runMany

    @PluginMethod
    fun runMany(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "runMany")
        val statement = call.getString("statement")
            ?: return failure(call, "INVALID_PARAMS", "'statement' is required", "runMany")
        if (statement.trim().isEmpty()) {
            return failure(call, "INVALID_PARAMS", "'statement' is required", "runMany")
        }
        val valueSets = try {
            val decoded = jsonArrayToValueList(call.getArray("values"), "values")
            if (decoded.isEmpty()) throw IllegalArgumentException("'values' must be a non-empty array of value arrays")
            decoded.mapIndexed { index, value ->
                @Suppress("UNCHECKED_CAST")
                value as? List<Any?>
                    ?: throw IllegalArgumentException("'values[$index]' must be an array")
            }
        } catch (e: IllegalArgumentException) {
            return failure(call, "INVALID_PARAMS", e.message ?: "Invalid values", "runMany")
        }
        val transaction = call.getBoolean("transaction", true) ?: true
        val returnResults = call.getBoolean("returnResults", false) ?: false

        executeSqlite {
            try {
                val result = impl.runMany(database, statement, valueSets, transaction, returnResults)
                val data = JSObject().put("changes", result.changes).put("lastInsertId", 0)
                result.results?.let { items ->
                    val encoded = JSArray()
                    items.forEach { item ->
                        encoded.put(JSObject().put("changes", item.changes).put("lastInsertId", item.lastInsertId))
                    }
                    data.put("results", encoded)
                }
                success(call, data)
            } catch (e: Exception) {
                failure(call, errorCode(e, "EXECUTE_FAILED"), e.message ?: "runMany failed", "runMany")
            }
        }
    }

    // MARK: - query

    @PluginMethod
    fun query(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "query")
        val statement = call.getString("statement")
            ?: return failure(call, "INVALID_PARAMS", "'statement' is required", "query")
        if (statement.trim().isEmpty()) {
            return failure(call, "INVALID_PARAMS", "'statement' is required", "query")
        }
        val values = try {
            jsonArrayToValueList(call.getArray("values"), "values")
        } catch (e: IllegalArgumentException) {
            return failure(call, "INVALID_PARAMS", e.message ?: "Invalid values", "query")
        }

        executeSqlite {
            try {
                if (call.getBoolean("__capacitorSqliteCompactRows", false) == true) {
                    val compact = impl.queryCompact(database, statement, values)
                    val encodedRows = JSArray()
                    compact.values.forEach { row ->
                        val encoded = JSArray()
                        row.forEach { value -> encoded.put(encodeQueryValue(value)) }
                        encodedRows.put(encoded)
                    }
                    success(
                        call,
                        JSObject().put(
                            "compactRows",
                            JSObject().put("columns", JSArray(compact.columns)).put("values", encodedRows)
                        )
                    )
                    return@executeSqlite
                }
                val rows = impl.query(database, statement, values)
                val result = JSArray()
                for (row in rows) {
                    val obj = JSObject()
                    for ((key, value) in row) {
                        obj.put(key, encodeQueryValue(value))
                    }
                    result.put(obj)
                }
                success(call, JSObject().put("rows", result))
            } catch (e: Exception) {
                failure(call, errorCode(e, "QUERY_FAILED"), e.message ?: "query failed", "query")
            }
        }
    }

    // MARK: - beginTransaction

    @PluginMethod
    fun beginTransaction(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "beginTransaction")

        executeSqlite {
            try {
                impl.beginTransaction(database)
                success(call)
            } catch (e: Exception) {
                failure(call, errorCode(e, "TRANSACTION_FAILED"), e.message ?: "beginTransaction failed", "beginTransaction")
            }
        }
    }

    // MARK: - commitTransaction

    @PluginMethod
    fun commitTransaction(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "commitTransaction")

        executeSqlite {
            try {
                impl.commitTransaction(database)
                success(call)
            } catch (e: Exception) {
                failure(call, errorCode(e, "TRANSACTION_FAILED"), e.message ?: "commitTransaction failed", "commitTransaction")
            }
        }
    }

    // MARK: - rollbackTransaction

    @PluginMethod
    fun rollbackTransaction(call: PluginCall) {
        val database = call.getString("database")
            ?: return failure(call, "INVALID_PARAMS", "'database' is required", "rollbackTransaction")

        executeSqlite {
            try {
                impl.rollbackTransaction(database)
                success(call)
            } catch (e: Exception) {
                failure(call, errorCode(e, "TRANSACTION_FAILED"), e.message ?: "rollbackTransaction failed", "rollbackTransaction")
            }
        }
    }

    // MARK: - JSON bridge helpers

    private fun jsonArrayToValueList(arr: JSArray?, label: String): List<Any?> {
        arr ?: return emptyList()
        return (0 until arr.length()).map { i ->
            val v = arr.get(i)
            unwrapJsonValue(v, "$label[$i]")
        }
    }

    private fun jsonArrayToStringList(arr: JSArray?, label: String): List<String> {
        arr ?: return emptyList()
        return (0 until arr.length()).map { i ->
            val value = arr.get(i) as? String
                ?: throw IllegalArgumentException("'$label[$i]' must be a string")
            require(value.trim().isNotEmpty()) { "'$label[$i]' must be a non-empty string" }
            value
        }
    }

    private fun jsonArrayToListOfMaps(arr: JSArray?, label: String): List<Map<String, Any?>> {
        arr ?: return emptyList()
        return (0 until arr.length()).map { i ->
            val obj = try {
                arr.getJSONObject(i)
            } catch (_: Exception) {
                throw IllegalArgumentException("'$label[$i]' must be an object")
            }
            val map = mutableMapOf<String, Any?>()
            val keys = obj.keys()
            while (keys.hasNext()) {
                val key = keys.next()
                val v = obj.get(key)
                map[key] = unwrapJsonValue(v, "$label[$i].$key")
            }
            map
        }
    }

    private fun unwrapJsonValue(v: Any, label: String): Any? = when (v) {
        JSONObject.NULL -> null
        is JSONArray    -> (0 until v.length()).map { i ->
            val item = v.get(i)
            unwrapJsonValue(item, "$label[$i]")
        }
        is JSONObject   -> decodeTaggedBlob(v, label)
        else -> v
    }

    /** Decode the compact internal BLOB envelope emitted by src/bridge-values.ts. */
    private fun decodeTaggedBlob(value: JSONObject, label: String): ByteArray {
        if (value.length() != 1 || !value.has(NATIVE_BLOB_BASE64_KEY)) {
            throw IllegalArgumentException("'$label' must not be an object")
        }
        val encoded = value.opt(NATIVE_BLOB_BASE64_KEY) as? String
            ?: throw IllegalArgumentException("'$label.$NATIVE_BLOB_BASE64_KEY' must be a base64 string")
        return try {
            Base64.decode(encoded, Base64.DEFAULT)
        } catch (_: IllegalArgumentException) {
            throw IllegalArgumentException("'$label.$NATIVE_BLOB_BASE64_KEY' must be valid base64")
        }
    }

    private fun encodeText(value: String): String {
        if (!value.startsWith(SQLiteHelpers.BLOB_PREFIX) && !value.startsWith(SQLiteHelpers.TEXT_PREFIX)) {
            return value
        }
        val encoded = Base64.encodeToString(value.toByteArray(StandardCharsets.UTF_8), Base64.NO_WRAP)
        return SQLiteHelpers.TEXT_PREFIX + encoded
    }

    private fun encodeQueryValue(value: Any?): Any = when (value) {
        null -> JSONObject.NULL
        is ByteArray -> SQLiteHelpers.BLOB_PREFIX + Base64.encodeToString(value, Base64.NO_WRAP)
        is String -> encodeText(value)
        else -> value
    }
}
